pkg/server/subresource/subresource.go

```diff
@@ -0,0 +1,129 @@
+	VirtualMachines        = schema.GroupVersionResource{Group: apiService.Group, Version: apiService.Version, Resource: "virtualmachines"}
+	PersistentVolumeClaims = schema.GroupVersionResource{Group: apiService.Group, Version: apiService.Version, Resource: "persistentvolumeclaims"}
+	VolumeSnapshots        = schema.GroupVersionResource{Group: apiService.Group, Version: apiService.Version, Resource: "snapshot.storage.k8s.io.volumesnapshot"}
+
+	namespacedSubResources    = []schema.GroupVersionResource{VirtualMachineImages, UpgradeLogs, VirtualMachines, PersistentVolumeClaims, VolumeSnapshots}
+	nonNamespacedSubResources = []schema.GroupVersionResource{Node}
+)
+
+type Resource struct {
+	Name        string
+	ObjectName  string
+	SubResource string
+	Namespace   string
+	Namespaced  bool
+}
+
+type ResourceHandler interface {
+	// IsMatchedResource checks if the resource, subresource and http method are matched
+	IsMatchedResource(resource Resource, method string) bool
+
+	// SubResourceHandler handles the request if `IsMatchedResource` is true.
+	SubResourceHandler(rw http.ResponseWriter, r *http.Request, resource Resource) error
+}
+
+var (
+	handlers        []ResourceHandler
+	healthCheckPath = fmt.Sprintf("/apis/%s/%s", apiService.Group, apiService.Version)
+)
+
+func injectResource(next http.Handler, resource string) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		mux.Vars(req)["resource"] = resource
+		next.ServeHTTP(rw, req)
+	})
+}
+
+func NewSubResourceHandler(mux *mux.Router) {
+	// force full matched routers instead of matching non-existed resource
+	for _, resource := range namespacedSubResources {
+		path := fmt.Sprintf("/apis/%s/%s/namespaces/{namespace}/%s/{name}/{subresource}", resource.Group, resource.Version, resource.Resource)
+		mux.Path(path).Handler(injectResource(http.HandlerFunc(serveSubResource), resource.Resource))
+	}
+
+	for _, resource := range nonNamespacedSubResources {
+		path := fmt.Sprintf("/apis/%s/%s/%s/{name}/{subresource}", resource.Group, resource.Version, resource.Resource)
+		mux.Path(path).Handler(injectResource(http.HandlerFunc(serveSubResource), resource.Resource))
+	}
+
+	mux.Path(healthCheckPath).Handler(http.HandlerFunc(healthCheck))
+}
+
+func RegisterSubResourceHandler(handler ResourceHandler) {
+	handlers = append(handlers, handler)
+}
+
+func healthCheck(rw http.ResponseWriter, _ *http.Request) {
+	rw.WriteHeader(http.StatusOK)
+}
+
+func serveSubResource(rw http.ResponseWriter, req *http.Request) {
+	vars := mux.Vars(req)
+
+	var (
+		err      error
+		found    bool
+		resource = Resource{
+			Name:        vars["resource"],
+			ObjectName:  vars["name"],
+			SubResource: vars["subresource"],
+			Namespace:   vars["namespace"],
+		}
+	)
+
+	if vars["namespace"] != "" {
+		resource.Namespaced = true
+		resource.Namespace = vars["namespace"]
+	}
+
+	logrus.Debug("Resource: ", resource)
```

Comment:
I think it would be a good idea to log any subresource requests. This would help tremendously when dealing with support cases as the service logs would then contain information on what actions happened and when.